### PR TITLE
Intra doc enum variant field

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of #43466.

Add intra-doc link support for this:

```rust
enum Foo {
    X {
        y: u8, // can be found with Foo::X::y
    }
}
```

r? @kinnison

### CODE REVIEW

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 1]
```diff
@@ -51,6 +51,59 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
...
+        if let Res::Err = ty_res {
+            return Err(());
+        }
+        let ty_res = ty_res.map_id(|_| panic!("unexpected node_id"));
+        match ty_res {
+            Res::Def(DefKind::Enum, did) => {
+                let item = cx.tcx.inherent_impls(did)
+                                 .iter()
+                                 .flat_map(|imp| cx.tcx.associated_items(*imp))
+                                 .find(|item| item.ident.name == variant_name);
```
Comment (lzutao) [inline src/librustdoc/passes/collect_intra_doc_links.rs]: `iter.find().is_some()` could be written as `iter.any()`

File: src/librustdoc/passes/collect_intra_doc_links.rs [hunk 2]
```diff
@@ -51,6 +51,59 @@ impl<'a, 'tcx> LinkCollector<'a, 'tcx> {
...
+                if item.is_some() {
+                    return Err(());
+                }
+                match cx.tcx.type_of(did).kind {
+                    ty::Adt(def, _) if def.is_enum() => {
+                        if def.all_fields()
+                              .find(|item| item.ident.name == variant_field_name).is_some() {
```
Comment (lzutao) [inline src/librustdoc/passes/collect_intra_doc_links.rs]: Here too: use `iter.any()`

**Review by kinnison — COMMENTED** [review]
Modulo the comments from @lzutao I think this looks good.

**Comment by GuillaumeGomez** [comment]
Updated.

**Review by kinnison — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Updated!

**Comment by GuillaumeGomez** [comment]
Ok, fixed the invalid errors return. We can move forward...

@bors: r=kinnison

**Pull request merged**